subsys/net/lib/http/http_hpack.c

```diff
@@ -0,0 +1,159 @@
+	[HTTP_HPACK_AGE] = (void *)"age",
+	[HTTP_HPACK_ALLOW] = (void *)"allow",
+	[HTTP_HPACK_AUTHORIZATION] = (void *)"authorization",
+	[HTTP_HPACK_CACHE_CONTROL] = (void *)"cache-control",
+	[HTTP_HPACK_CONTENT_DISPOSITION] = (void *)"content-disposition",
+	[HTTP_HPACK_CONTENT_ENCODING] = (void *)"content-encoding",
+	[HTTP_HPACK_CONTENT_LANGUAGE] = (void *)"content-language",
+	[HTTP_HPACK_CONTENT_LENGTH] = (void *)"content-length",
+	[HTTP_HPACK_CONTENT_LOCATION] = (void *)"content-location",
+	[HTTP_HPACK_CONTENT_RANGE] = (void *)"content-range",
+	[HTTP_HPACK_CONTENT_TYPE] = (void *)"content-type",
+	[HTTP_HPACK_COOKIE] = (void *)"cookie",
+	[HTTP_HPACK_DATE] = (void *)"date",
+	[HTTP_HPACK_ETAG] = (void *)"etag",
+	[HTTP_HPACK_EXPECT] = (void *)"expect",
+	[HTTP_HPACK_EXPIRES] = (void *)"expires",
+	[HTTP_HPACK_FROM] = (void *)"from",
+	[HTTP_HPACK_HOST] = (void *)"host",
+	[HTTP_HPACK_IF_MATCH] = (void *)"if-match",
+	[HTTP_HPACK_IF_MODIFIED_SINCE] = (void *)"if-modified-since",
+	[HTTP_HPACK_IF_NONE_MATCH] = (void *)"if-none-match",
+	[HTTP_HPACK_IF_RANGE] = (void *)"if-range",
+	[HTTP_HPACK_IF_UNMODIFIED_SINCE] = (void *)"if-unmodified-since",
+	[HTTP_HPACK_LAST_MODIFIED] = (void *)"last-modified",
+	[HTTP_HPACK_LINK] = (void *)"link",
+	[HTTP_HPACK_LOCATION] = (void *)"location",
+	[HTTP_HPACK_MAX_FORWARDS] = (void *)"max-forwards",
+	[HTTP_HPACK_PROXY_AUTHENTICATE] = (void *)"proxy-authenticate",
+	[HTTP_HPACK_PROXY_AUTHORIZATION] = (void *)"proxy-authorization",
+	[HTTP_HPACK_RANGE] = (void *)"range",
+	[HTTP_HPACK_REFERER] = (void *)"referer",
+	[HTTP_HPACK_REFRESH] = (void *)"refresh",
+	[HTTP_HPACK_RETRY_AFTER] = (void *)"retry-after",
+	[HTTP_HPACK_SERVER] = (void *)"server",
+	[HTTP_HPACK_SET_COOKIE] = (void *)"set-cookie",
+	[HTTP_HPACK_STRICT_TRANSPORT_SECURITY] = (void *)"strict-transport-security",
+	[HTTP_HPACK_TRANSFER_ENCODING] = (void *)"transfer-encoding",
+	[HTTP_HPACK_USER_AGENT] = (void *)"user-agent",
+	[HTTP_HPACK_VARY] = (void *)"vary",
+	[HTTP_HPACK_VIA] = (void *)"via",
+	[HTTP_HPACK_WWW_AUTHENTICATE] = (void *)"www-authenticate",
+};
+
+int http_hpack_table_contains(struct http_client_ctx *server, uint32_t key)
+{
+	if (!server || !http_hpack_key_is_static((int32_t)key)) {
+		return -EINVAL;
+	}
+
+	if (http_hpack_table_static[key]) {
+		return 0;
+	} else {
+		return -EINVAL;
+	}
+}
+
+int http_hpack_table_get(struct http_client_ctx *server, uint32_t key, void **value)
+{
+	if (!server || !value || !http_hpack_key_is_static((int32_t)key)) {
+		return -EINVAL;
+	}
+
+	*value = (void *)http_hpack_table_static[key];
+	if (*value) {
+		return 0;
+	} else {
+		return -EINVAL;
+	}
+}
+
+int http_hpack_table_add(struct http_client_ctx *ctx, uint32_t key, void *value)
+{
+	return -ENOSYS;
+}
+
+int http_hpack_table_remove(struct http_client_ctx *ctx, uint32_t key)
+{
+	return -ENOSYS;
+}
+
+char *header_parsing(struct http_client_ctx *ctx, int requested_data)
```

Comment:
This function needs a considerable design change. We should not be "requesting" anything from the parsing function. Instead, we need to provide it with buffers for header name and value, and let it fill it out with content.

I suggest that a single function call parse a single HPACK header entry (therefore it'd make sense to call it `http_hpack_parse_header()`).

Parsing HPACK entry is a bit trickier than parsing frame header from insteance, as we don't know beforehand what length to expect, so you need to always check during the parsing whether there is enough data in the buffer to proceed (for example if you read Literal Header "Name Length" field, and it would indicate that the actual string is longer than the remaining data in the buffer, you can return `-EAGAIN` and start over from scratch on the next loop iteration (when more data arrives).

I think that for simplicity, we can assume that we only support HPACK headers of size lower than `CONFIG_NET_HTTP_CLIENT_BUFFER_SIZE` - otherwise it'll become much more complex task to provide a HPACK parser able to deal with partial (incomplete) headers. IMO, for embedded system this is a reasonable tradeoff, I doubt we'll have a need to support very long header fields any time soon (like cookies).

And once the function finally returns a success - it should be the caller who decides what to do with the new data. If it's a header of interest for us (like URL, Method etc), store it. If it's not, just ignore it and proceed to the next entry.